File: apps/java/src/test/java/chapter15/FileModelStoreTest.java
package chapter15;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.Assertions.within;

import chapter02.Row;
import chapter02.Table;
import chapter07.LinearModel;
import chapter08.ClassWeight;
import chapter08.Pipeline;
import chapter08.SurvivedData;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.OptionalDouble;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class FileModelStoreTest {
  @TempDir Path directory;

  private static Passenger passenger(String sex) {
    return new Passenger(2, sex, OptionalDouble.empty(), 0, 0, 20.0, Optional.empty());
  }

  /** 女性は生存、男性は死亡という架空の乗客 8 人。 */
  private static Table fictionalPassengers() {
    String[][] values = {
      {"1", "female", "25", "0", "0", "60", "C"},
      {"2", "female", "35", "1", "0", "30", "S"},
      {"3", "female", "18", "0", "1", "10", "Q"},
      {"3", "female", "", "0", "0", "9", "S"},
      {"1", "male", "40", "1", "0", "55", "C"},
      {"2", "male", "28", "0", "0", "15", "S"},
      {"3", "male", "22", "0", "0", "8", ""},
      {"3", "male", "", "0", "0", "7", "S"},
    };
    List<Row> rows = new ArrayList<>();
    for (String[] row : values) {
      Map<String, String> cells = new HashMap<>();
      for (int i = 0; i < row.length; i++) {
        cells.put(SurvivedData.FEATURES.get(i), row[i]);
      }
      rows.add(new Row(cells));
    }
    return new Table(SurvivedData.FEATURES, rows);
  }

  @Test
  @DisplayName("保存した線形回帰モデルを読み込んで興行収入を予測する")
  void salesModel() throws Exception {
    var store = new FileModelStore(directory);
    store.saveSalesModel(
        new LinearModel(100.0, Map.of("SNS1", 1.0, "SNS2", 2.0, "actor", 0.5, "original", 10.0)));

    SalesModel model = store.loadSalesModel();

    assertThat(model.predictSales(new Movie(10.0, 20.0, 100.0, 1))).isCloseTo(210.0, within(1e-9));
  }

  @Test
  @DisplayName("保存したパイプラインを読み込んで生存を判定する")
  void survivalModel() throws Exception {
    var store = new FileModelStore(directory);
    store.saveSurvivalModel(
        Pipeline.build(2, ClassWeight.NONE)
            .fit(fictionalPassengers(), List.of(1, 1, 1, 1, 0, 0, 0, 0)));

    SurvivalModel model = store.loadSurvivalModel();

    assertThat(model.survives(passenger("female"))).isTrue();
    assertThat(model.survives(passenger("male"))).isFalse();
  }

  @Test
  @DisplayName("モデルファイルが無ければ、パスを含まない ModelNotFoundException を投げる")
  void missingFiles() {
    var store = new FileModelStore(directory);

    assertThatThrownBy(store::loadSalesModel)
        .isInstanceOf(ModelNotFoundException.class)
        .hasMessageNotContaining(directory.toString());
    assertThatThrownBy(store::loadSurvivalModel)
        .isInstanceOfSatisfying(
            ModelNotFoundException.class, e -> assertThat(e.model()).isEqualTo("survived"));
  }
}

File: apps/fsharp/tests/MachineLearning.Tests/Chapter15/FileModelStoreTest.fs
module MachineLearning.Tests.Chapter15.FileModelStoreTest

open System.IO
open Xunit
open MachineLearning.Chapter07.LinearRegression
open MachineLearning.Chapter08.Pipeline
open MachineLearning.Chapter08.WeightedTree
open MachineLearning.Chapter15.Domain
open MachineLearning.Chapter15.FileModelStore
open MachineLearning.Tests.Chapter15.Stubs

/// テストごとに空の一時ディレクトリを作る
let emptyDirectory () : string =
    Directory.CreateTempSubdirectory("model-").FullName

let salesModel: LinearModel =
    {
        Intercept = 100.0
        Coefficients = Map.ofList [ "SNS1", 2.0; "SNS2", 0.5; "actor", 1.0; "original", 10.0 ]
    }

[<Fact>]
let ``保存した線形回帰を読み込み、映画の特徴量から興行収入を予測する`` () =
    let directory = emptyDirectory ()
    saveSalesModel directory salesModel

    match (fileModelStore directory).LoadSalesModel() with
    | Ok predictSales ->
        // 100 + 2 × 200 + 0.5 × 500 + 1 × 3000 + 10 × 1（原作あり）
        Assert.Equal(3760.0, predictSales movie, 9)
    | Error error -> Assert.Fail $"%A{error}"

[<Fact>]
let ``モデルのファイルが無ければ ModelNotFound を返す`` () =
    let store = fileModelStore (emptyDirectory ())

    Assert.Equal(Error(ModelNotFound "cinema"), store.LoadSalesModel() |> Result.map ignore)
    Assert.Equal(Error(ModelNotFound "survived"), store.LoadSurvivalModel() |> Result.map ignore)

[<Fact>]
let ``モデルのファイルが壊れていれば ModelUnreadable を返す`` () =
    let directory = emptyDirectory ()
    File.WriteAllText(Path.Combine(directory, SalesModelFile), "{")

    Assert.Equal(Error(ModelUnreadable "cinema"), (fileModelStore directory).LoadSalesModel() |> Result.map ignore)

[<Fact>]
let ``ドメインの乗客を第 8 章の乗客に変換する`` () =
    let converted = toChapter08Passenger passenger

    Assert.Equal(1, converted.Pclass)
    Assert.Equal("female", converted.Sex)
    Assert.Equal(None, converted.Age)
    Assert.Equal(Some "C", converted.Embarked)

[<Fact>]
let ``保存したパイプラインを読み込み、第 8 章と同じ生存の予測をする`` () =
    let passengers =
        [ Male, 30.0; Female, 20.0; Male, 60.0; Female, 40.0 ]
        |> List.map (fun (sex, age) ->
            toChapter08Passenger
                { passenger with
                    Sex = sex
                    Age = Some age
                })

    let pipeline =
        fitPipeline
            {
                MaxDepth = Some 2
                ClassWeight = Unweighted
            }
            passengers
            [ 0; 1; 0; 1 ]

    let directory = emptyDirectory ()
    saveSurvivalModel directory pipeline

    match (fileModelStore directory).LoadSurvivalModel() with
    | Ok survives ->
        for p in [ passenger; { passenger with Sex = Male } ] do
            Assert.Equal(predictPassenger pipeline (toChapter08Passenger p) = 1, survives p)
    | Error error -> Assert.Fail $"%A{error}"
